### DISCUSSION TITLE
AWS python lambda - fails to import oracledb

### DISCUSSION THREAD

**Discussion by ralphwillgoss** [oracle/python-oracledb#32] [Q&A] [answered]
I'm trying to get the new oracledb library working in a AWS python lambda, using python 3.9.

After following the basic installation procedure and deploying my lambda, I get the following error:
_{
  "errorMessage": "Unable to import module 'main': cannot import name 'base_impl' from partially initialized module 'oracledb' (most likely due to a circular import) (/var/task/oracledb/__init__.py)",
  "errorType": "Runtime.ImportModuleError",
   {snip}
}_

I've installed the library as per the instructions and created a simple zip file lambda deployment.

My lambda directory structure looks like the following:
![image](https://user-images.githubusercontent.com/1279021/178308888-7eab6352-434d-40a0-aab3-61f8da7423d0.png)

I'm using a simple entry point to test the import of the library, main.py looks like:
```
import oracledb

def lambda_handler(event, context):
    pass
```

This all works fine locally, is there something subtle I have missed?

**Comment by anthony-tuininga** [comment, marked as answer]
I don't know much about AWS lambda but the error you are getting is suggesting that the shared library (C extension) is not being loaded properly. My directory looks like this:

```
aq.py
base_impl.cpython-310-x86_64-linux-gnu.so
connection.py
connect_params.py
constants.py
constructors.py
cursor.py
dbobject.py
defaults.py
driver_mode.py
dsn.py
errors.py
exceptions.py
__init__.py
lob.py
pool_params.py
pool.py
py.typed
soda.py
subscr.py
thick_impl.cpython-310-x86_64-linux-gnu.so
thin_impl.cpython-310-x86_64-linux-gnu.so
utils.py
var.py
version.py
```

Take note of the 3 files with .so -- these are the ones that your installation doesn't seem to be able to find!

**Reply by ralphwillgoss** [reply]
thanks @anthony-tuininga , I'll look more closely at the build server and update the discussion.

**Reply by ralphwillgoss** [reply]
@anthony-tuininga thanks for tip, that was the issue and once the package installed successfully all worked.
Thanks again!

**Reply by anthony-tuininga** [reply]
You're welcome!

**Reply by cbackas** [reply]
> @anthony-tuininga thanks for tip, that was the issue and once the package installed successfully all worked. Thanks again!

Also running into this, what was the problem with your build server configuration? ~~I'm looking in the oracledb folder that my build server produces and it has the *.cpython-310-x86_64-linux-gnu.so files there.~~ Oh I think the build server is using python 3.7 and the lambda is configured to use 3.9 - will try aligning those

**Reply by ralphwillgoss** [reply]
@cbackas let us know how you go
- Ensure you not building on a Windows box if your build setup can select from a pool of build nodes to build on
- Ensure there are no file filters on the steps that create the artefacts for deployment

**Reply by cbackas** [reply]
> @cbackas let us know how you go
> 
> - Ensure you not building on a Windows box if your build setup can select from a pool of build nodes to build on
> 
> - Ensure there are no file filters on the steps that create the artefacts for deployment

Yeah do have a Linux build server so it was indeed the python version. While it would have been nice to use 3.9 the path of least resistance was to just lower the lambda environment to 3.7 and now I'm onto debugging other problems :)

**Reply by ralphwillgoss** [reply]
@cbackas I just checked and I'm using `Python 3.9` and `x86_64` and all works well.

**Reply by cbackas** [reply]
> @cbackas I just checked and I'm using `Python 3.9` and `x86_64` and all works well.

Oh yeah not saying 3.9 wouldn't work perfectly fine, just that I didn't want to change the python version our Jenkins nodes use :P

**Reply by ThePhoenixRising70** [reply]
I'm struggling with this issue and I'm unclear what the directory structure needs to be.  I'm trying to put this in a Lambda Layer so it can be maintained when new releases of python-oracledb occur. I'm building the zip file in an Amazon Linux 2 EC2.
It is unclear for me from reading this thread what the directory structure needs to be to include all the dependencies (cffi, cryptography, etc.) and be able to be accessible from the lambda function.  Will greatly appreciate any assistance.

**Reply by cbackas** [reply]
@ThePhoenixRising70 

You want to install your pip packages to the same place as your code:
`python3 -m pip install -r requirements.txt -t .` the `-t .` would place the packages in your current directory. 

Here's the inside of one of the versions of my lambda zip. I highlighted the 3 files that have my actual python code, everything else came from PIP install.
<img width="386" alt="image" src="https://user-images.githubusercontent.com/4987637/182458268-42191f6d-7c33-4c5a-834d-261700782cca.png">

Can find step by step here: https://docs.aws.amazon.com/lambda/latest/dg/python-package.html#python-package-create-package-with-dependency

**Reply by ralphwillgoss** [reply]
@ThePhoenixRising70 your zip file should represent the directory structure I have, in my screen shot in the question.

Consider the directory `ralph-oracle-test` as being the container you should zip and then any directories (e.g. oracledb), which are below this - are now in the root of you zip and hence your lambda too.

Any new packages I would install, I would target their installation into the `ralph-oracle-test` directory, as this will always be the root of my zip.

**Reply by ghost** [reply]
@ThePhoenixRising70  did you ever get your Lambda layer working? could you share how you got it working?
I created a layer and I'm currently getting:
python-oracledb thin mode cannot be used because the cryptography package is not installed

I am not sure what I am missing. I installed my requirements.txt into /my_layer_root/python
![image](https://user-images.githubusercontent.com/16198534/192900269-886619eb-2555-4771-b84c-e90c2e070952.png)
Inside the utils folder I got my custom code that uses the oracledb. 
what am I missing for the layer config?

**Reply by ralphwillgoss** [reply]
@github4es have you set the appropriate environment variable, so your lambda picks up the layer? does the image represent what is the root of the zip file?

**Reply by ghost** [reply]
@ralphwillgoss thanks for reaching out. My lambda is picking up the layer correctly since it is able to execute the `import oracledb` and the line that fails which is `oracledb.connect( ... )` - otherwise, it wouldn't be able to import the oracledb package.

**Reply by ralphwillgoss** [reply]
@github4es ahh, thanks Ed. I've had some similar errors before - just need to jog my memory as to why they occurred.

**Reply by ghost** [reply]
We're currently using the cx_oracle in a layer but is pretty massive. Were really excited to try `oracledb` because it is so much smaller.

**Reply by ybindu** [reply]
Hi @github4es, did you manage to resolve your issue? 'python-oracledb thin mode cannot be used because the cryptography package is not installed'

I'm facing the same issue. Even though I got crypto packages in my root folder, I'm getting this error.

<img width="448" alt="image" src="https://user-images.githubusercontent.com/108888177/195698812-1b5ebbc4-f2b6-4fb4-bdec-a5071f4adc55.png">

Could someone guide me to the solution

**Reply by anthony-tuininga** [reply]
That error occurs because the cryptography package cannot be imported. Try these commands in your main script:

```python
from cryptography import x509
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.ciphers import algorithms, modes, Cipher
from cryptography.hazmat.primitives.asymmetric import padding
from cryptography.hazmat.primitives.kdf import pbkdf2
```

That should help determine the source of the problem.

**Reply by shivanibhavsar02** [reply]
Hi @anthony-tuininga , i am also facing same issue, the oracledb directory screenshot which you have put has 3 main files as you said, those are with .so extension . But when i installed oracledb in my local, it has those 3 files with .pyd extension. Is there any issue with that?

**Reply by anthony-tuininga** [reply]
The extension ``.so`` is used on Linux. On Windows the extension is ``.pyd``. That isn't a problem.

**Reply by ralphwillgoss** [reply]
@shivanibhavsar02 as mentioned [below](https://github.com/oracle/python-oracledb/discussions/32#discussioncomment-8394113), the `.pyd` extensions are Windows specific. If you try and install this packaged lambda with Windows binaries on an AWS, which is running Linux, you will get the import errors mentioned in this article.

**Comment by dharanispg** [comment]
Hello Guys, Same issue with AWS Glue as well , but I have confirmed that .so files existence in the zip. still it's failing 
**ImportError: cannot import name 'base_impl' from 'oracledb' (/tmp/modules.zip/oracledb/__init__.py)** 
Any reasons why ?

**Reply by anthony-tuininga** [reply]
Looks like you are trying to load the module directly from the zip file. This works fine for pure Python files but will fail with extension modules (like base_impl). You'll need to make sure that zip file is extracted on disk.

**Reply by peanutbutter97** [reply]
i zipped the the folder structure exactly like you did but i still get the error why? 
```
{
"errorMessage": "Unable to import module 'main': cannot import name 'base_impl' from partially initialized module 'oracledb' (most likely due to a circular import) (/var/task/oracledb/init.py)",
"errorType": "Runtime.ImportModuleError",
{snip}
}
```

**Reply by anthony-tuininga** [reply]
The files need to be directly on disk to be loaded. They cannot be zipped up.

**Reply by peanutbutter97** [reply]
@anthony-tuininga hi thanks for reply. How to make sure it don’t zip on aws lambda when I create a layer?

**Reply by anthony-tuininga** [reply]
I don't know how AWS lambda works precisely so can't really help with that. Some of the others here may be able to help, though!

**Reply by cjbj** [reply]
There are various Google search hits for cx_Oracle and Lambda which won't be too different from the steps needed for python-oracledb.

**Reply by ETzaf** [reply]
> The files need to be directly on disk to be loaded. They cannot be zipped up.

Hello, i am facing the same issue. Can you be for specific to your answer please? Thank you

**Reply by peanutbutter97** [reply]
Set up an EC2 to install the lib then zipped out.

**Reply by Takahir-O** [reply]
> > The files need to be directly on disk to be loaded. They cannot be zipped up.
> 
> Hello, i am facing the same issue. Can you be for specific to your answer please? Thank you

Hello. I had the same situation and solved it in the following way. (I do not use EC2)
(1) I matched the local python version with the Lambda version.
The local python version is 3.10, but Lambda is python 3.9, so I set it to 3.9.

(2) I downloaded the Linux-compatible version to a directory by pip, as shown in this link
https://aws.amazon.com/jp/premiumsupport/knowledge-center/lambda-python-package-compatible/

Note that if your local python is 3.10, the module downloaded by this method will be the 3.10 version.

Finally, I added the downloaded folder to the Lambda layer.

Hope it works. Thank you.

**Reply by cjbj** [reply]
Thanks for sharing this info.

**Reply by antonizero99** [reply]
Thanks @Takahir-O . You saved my day.

**Comment by RabiaHasanZaheer** [comment]
I'm getting the same error in AWS Lambda and I've check library is correctly installed. Also installed cryptography just in case. I installed linux compatible version.

**Reply by ralphwillgoss** [reply]
how did you create the lambda zip deployment package?

**Reply by shivanibhavsar02** [reply]
@ralphwillgoss Hi, i am also facing same issue, i have installed all libraries correctly. i am running the code in my local in windows OS and in AWS lambda. At both places i am facing same issue. Can you help please?

**Reply by ralphwillgoss** [reply]
@shivanibhavsar02 you need to ensure that the system you are deploying from, matches the system you are deploying too
Windows -> Windows
Linux -> Linux

In your case, you cannot deploy from Windows -> AWS (Linux), as the Oracle binaries would have been packaged according to the the system deploying (Windows).

**Comment by tsalyers** [comment]
I had a very similar problem to this. At first it was because I downloaded the library on a Windows machine before I zipped it up, but it wasn't working even with a Linux version. 

Then I tried downloading it with the binary-only flag (`pip install --only-binary=:all: oracledb -t .`) and that seems to have worked. I'm able to connect to Oracle from my test lambda function.

**Reply by ralphwillgoss** [reply]
thanks @tsalyers, glad this helped

**Comment by AjGitHubStore** [comment]
> Hello Guys, Same issue with AWS Glue as well , but I have confirmed that .so files existence in the zip. still it's failing **ImportError: cannot import name 'base_impl' from 'oracledb' (/tmp/modules.zip/oracledb/init.py)** Any reasons why ?

Are you able to resolve an issue? I am facing the same issue. Glue version 4, Python version 3.10

**Reply by vignxs** [reply]
You can use additional python module option to install oracledb . This option is in job parameters

**Reply by AjGitHubStore** [reply]
That's not an option for me ... I am going with the zip file option.
